### DISCUSSION TITLE
Best way to deal with arrays that should be empty but are instead omitted

### DISCUSSION THREAD

**Discussion by mwmx** [colinhacks/zod#651] [Q&A]
I'm using zod with a backend framework that helpfully [\*] strips out empty arrays when generating json responses. (\* In fact, it isn't helpful at all.)

What I mean is that a response like `{ "boolfield": true, "arrayfield": [] }` is turned into `{ "boolfield": true }`.

Since dealing with zero-length arrays is so much easier, I've started using `z.array(z.typeinquestion).optional().transform((val) => { if (!val) { return []; } return val; })`

More concretely:

```typescript
import { z } from 'zod';

describe('zod exercises', () => {
    it('empty and non-present arrays coming from the server', () => {
        const testValidator = z.object({
            boolfield: z.boolean(),
            arrayfield: z
                .array(z.number())
                .optional()
                .transform((val) => {
                    if (!val) {
                        return [];
                    }
                    return val;
                }),
        });

        const out1 = testValidator.parse({
            boolfield: true,
            arrayfield: [1, 2, 3, 4],
        });
        const out2 = testValidator.parse({ boolfield: true, arrayfield: [] });
        const out3 = testValidator.parse({ boolfield: true });

        expect(out1.arrayfield.length).toBe(4);
        expect(out2.arrayfield.length).toBe(0);
        expect(out3.arrayfield.length).toBe(0);
    });
});
```

This works just as I hoped it would, but I wish I weren't all but forced into it 😄 
Is there a more concise way to achieve the same thing?

**Comment by mwmx** [comment]
Here's one possible solution I found:

```typescript
const coerceFalsyToArray = (arg: ZodTypeAny) => {
    return z
        .array(arg)
        .optional()
        .transform((val) => {
            if (!val) {
                return [];
            }
            return val;
        });
};

const testValidator = z.object({
    boolfield: z.boolean(),
    arrayfield: coerceFalsyToArray(z.number()),
});
```

One downside to this: if you use `z.infer` with this `arrayfield` is `any[]` which is not ideal. I guess you could work around it using `Omit` and `.extend` but that's not very nice.

Another downside: `coerceFalsyToArray` is a lousy name. 🤔

**Reply by colinhacks** [reply]
You should write a generic function if you want type inference to continue working as expected:

```ts
const coerceFalsyToArray = <T extends z.ZodTypeAny>(arg: T) => {
  return z
    .array(arg)
    .optional()
    .transform((val) => {
      if (!val) {
        return [];
      }
      return val;
    });

  
};
const mySchema = coerceFalsyToArray(z.string());
type mySchema = z.infer<typeof mySchema>; // string[]
```

Though imo your current solution is cleaner and more explicit. You can abbreviate it significantly with an inline arrow function and nullish coalescing:

```ts
const b = z
    .array(z.number())
    .optional()
    .transform((val) => val ?? []);
```

**Reply by mwmx** [reply]
@colinhacks thanks.

But why do you say that my solution is cleaner? The only difference is that yours uses generics and mine doesn't.

**Reply by mwmx** [reply]
Duh, you meant that the code in my original question was cleaner than `coerceFalsyToArray`. :)

I'm not sure I agree but at least I see what you were getting at.

**Comment by StefanTerdell** [comment]
Maybe .default() is what you're looking for? Passes the tests you provided :)

``` typescript
const testValidator = z.object({
  boolfield: z.boolean(),
  arrayfield: z
    .array(z.number()).default([])
});
```
